feat: add iter/while-each (#1180)
stdlib-js/stdlib: lib/node_modules/@stdlib/iter/while-each/README.md

r = it.next().value;
// undefined

// ...
```

The returned iterator protocol-compliant object has the following properties:

-   **next**: function which returns an iterator protocol-compliant object containing the next iterated value (if one exists) assigned to a `value` property and a `done` property having a boolean value indicating whether the iterator is finished.
-   **return**: function which closes an iterator and returns a single (optional) argument in an iterator protocol-compliant object.

Both the `predicate` function and the function to invoke for each iterated value are provided two arguments:

-   **value**: iterated value
-   **index**: iteration index (zero-based)

```javascript
var array2iterator = require( '@stdlib/array/to-iterator' );

function predicate( v ) {
    return v < 3;
}

function assert( v, i ) {
    if ( i > 1 ) {
        throw new Error( 'unexpected error' );
    }
}

var it = iterWhileEach( array2iterator( [ 1, 2, 3, 4 ] ), predicate, assert );
// returns <Object>

var r = it.next().value;
// returns 1

r = it.next().value;
// returns 2

r = it.next().value;
// undefined

// ...
```

To set the execution context for `fcn`, provide a `thisArg`.

> <!-- eslint-disable no-undefined-this -->

Reviewer: ```suggestion
<!-- eslint-disable no-invalid-this -->
```

<!--

@license Apache-2.0

Copyright (c) 2023 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

> # whileEach Iterator

Reviewer: ```suggestion
# iterWhileEach
```

> Create an iterator which, while a test condition is true, invokes a function for each iterated value before returning the iterated value.

<!-- Section to include introductory text. Make sure to keep an empty line after the intro `section` element and another before the `/section` close. -->

<section class="intro">

</section>

<!-- /.intro -->

<!-- Package usage documentation. -->

<section class="usage">

## Usage

```javascript
var iterWhileEach = require( '@stdlib/iter/while-each' );
```

#### iterWhileEach( iterator, predicate, fcn\[, thisArg] )

> Returns an iterator which invokes a `function` for each iterated value untill either a `predicate` returns `false` or the function has iterated over all value **before** returning the iterated value.

Reviewer: ```suggestion
Returns an iterator which invokes a function for each iterated value **before** returning the iterated value until either a `predicate` function returns `false` or the iterator has iterated over all values.
```

```javascript
var array2iterator = require( '@stdlib/array/to-iterator' );

function predicate( v ) {
    return v < 3;
}

function assert( v ) {
    if ( v !== v ) {
        throw new Error( 'should not be NaN' );
    }
}

var it = iterWhileEach( array2iterator( [ 1, 2, 3, 4 ] ), predicate, assert );
// returns {}

var r = it.next().value;
// returns 1

r = it.next().value;
// returns 2

r = it.next().value;
// undefined

// ...
```

The returned iterator protocol-compliant object has the following properties:

> -   **next**: function which returns an iterator protocol-compliant object containing the next iterated value (if one exists) assigned to a `value` property and a `done` property having a `boolean` value indicating whether the iterator is finished.

Reviewer: ```suggestion
-   **next**: function which returns an iterator protocol-compliant object containing the next iterated value (if one exists) assigned to a `value` property and a `done` property having a boolean value indicating whether the iterator is finished.
```

-   **return**: function which closes an iterator and returns a single (optional) argument in an iterator protocol-compliant object.

> The invoked `function` is provided two arguments:

Reviewer: ```suggestion
Both the `predicate` function and the function to invoke for each iterated value are provided two arguments:
```

> -   `value`: iterated value
> -   `index`: iteration index (zero-based)

Reviewer: ```suggestion
-   **value**: iterated value
-   **index**: iteration index (zero-based)
```

```javascript
var array2iterator = require( '@stdlib/array/to-iterator' );

function predicate( v ) {
    return v < 3;
}

> function assert( v ) {
>     if ( v !== v ) {
>         throw new Error( 'should not be NaN' );
>     }
> }

Reviewer: ```suggestion
function assert( v, i ) {
    if ( i > 1 ) {
        throw new Error( 'unexpected error' );
    }
}
```

I see that `iter/for-each` has a similar example, which does not actually illustrate anything new. Here, we at least want to show that provided functions are invoked with multiple arguments.

Reviewer: Once this PR is merged, feel free to submit a PR against `iter/for-each` which updates the respective example to be more illustrative.

var it = iterWhileEach( array2iterator( [ 1, 2, 3, 4 ] ), predicate, assert );
// returns <Object>

var r = it.next().value;
// returns 1

r = it.next().value;
// returns 2

r = it.next().value;
// undefined

// ...
```

> To set the function execution context, provide a `thisArg`.

Reviewer: ```suggestion
To set the execution context for `fcn`, provide a `thisArg`.
```

> ```javascript

Reviewer: ```suggestion
<!-- eslint-disable no-undefined-this -->

```javascript
```

I believe we need to disable this lint rule here, as I believe we should be getting a lint error.

var array2iterator = require( '@stdlib/array/to-iterator' );

function assert( v ) {
    this.count += 1;
    if ( v !== v ) {
        throw new Error( 'should not be NaN' );
    }
}

function predicate( v ) {
    return v < 3;
}

var ctx = {
    'count': 0
};

var it = iterWhileEach( array2iterator( [ 1, 2, 3 ] ), predicate, assert, ctx );
// returns <Object>

var r = it.next().value;
// returns 1

r = it.next().value;
// returns 2

var count = ctx.count;
// returns 2
```

</section>

<!-- /.usage -->

<!-- Package usage notes. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="notes">

## Notes

-   If an environment supports `Symbol.iterator` **and** a provided iterator is iterable, the returned iterator is iterable.

</section>

<!-- /.notes -->

<!-- Package usage examples. -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var randu = require( '@stdlib/random/iter/randu' );
var isnan = require( '@stdlib/math/base/assert/is-nan' );

> var iterForEach = require( '@stdlib/iter/for-each' );

Reviewer: ```suggestion
var iterWhileEach = require( '@stdlib/iter/while-each' );
```

function assert( v ) {
    if ( isnan( v ) ) {
        throw new Error( 'should not be NaN' );
    }
}

function predicate( v ) {

>     return v < 0.5;

Reviewer: ```suggestion
    return v <= 0.75;
```

Updating to match `examples/index.js`.

}

// Create a seeded iterator for generating pseudorandom numbers:
var rand = randu({
    'seed': 1234,
    'iter': 10
});

// Create an iterator which validates generated numbers:

> var it = iterForEach( rand, predicate, assert );

Reviewer: ```suggestion
var it = iterWhileEach( rand, predicate, assert );
```

// Perform manual iteration...
var r;
while ( true ) {
    r = it.next();
    if ( r.done ) {
        break;
    }
    console.log( r.value );
}
```

</section>

<!-- /.examples -->

<!-- Section to include cited references. If references are included, add a horizontal rule *before* the section. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="references">

</section>

<!-- /.references -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

> 
> * * *
> 
> ## See Also
> 
> -   <span class="package-name">[`@stdlib/iter/map`][@stdlib/iter/map]</span><span class="delimiter">: </span><span class="description">create an iterator which invokes a function for each iterated value.</span>

Reviewer: ```suggestion
```

We handle "See Also" sections in a separate build process which draws from a central project database.

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

> <!-- <related-links> -->
> 
> [@stdlib/iter/map]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/iter/map
> 
> <!-- </related-links> -->
> 

Reviewer: ```suggestion
```
